docs/proposals/third-party-api-servers.md

```diff
@@ -0,0 +1,175 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+<strong>
+The latest 1.0.x release of this document can be found
+[here](http://releases.k8s.io/release-1.0/docs/proposals/third-party-api-servers.md).
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Third Party API Servers
+
+## Goal
+
+With third party API servers, we want to enable cluster admins or users to
+expose new APIs at runtime without having to modify the core kubernetes API
+server binary.
+Anyone can write their own API server and register it with the core
+kubernetes API server. These new APIs will be seamless extension of core
+kubernetes APIs (ex: they will be operated upon via kubectl).
+
+## Non Goals:
+
+The following are related but are not the goals of this specific proposal:
+* Make it easy to write a kubernetes API server.
+
+## High Level Architecture
+
+Users can bring up their own API servers and register them with the core server
+at runtime.
+Core server will then start exposing these third party servers in the discovery
+API. kubectl will use the discovery API to connect to these servers.
+
+These API servers will have full control of their resources.
+Each server will have their own storage mechanism and will implement their own
+authorization and authentication.
+
+
+## Implementation Details
+
+### Thirdpartyapiservers Extensions Resource
+
+API server will support a non-namespaced `thirdpartyapiservers` extension resource.
+Third party servers can POST at
+`http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to register themselves.
+Clients can GET `http://kubernetes-master/apis/extensions/v1beta1/thirdpartyapiservers/`
+to get a list of registered third party servers.
+The API definition will be as follows:
+
+```
+// ThirdPartyAPIServer represents the configuration of a third party server.
+type ThirdPartyAPIServer struct {
+  unversioned.TypeMeta
+
+  ObjectMeta
+
+  Spec ThirdPartyAPIServerSpec
+
+  Status ThirdPartyAPIServerStatus
+}
+
+// ThirdPartyAPIServerSpec is the specification of a ThirdPartyAPIServer
+// resource.
+type ThirdPartyAPIServerSpec struct {
+  // Base path of the API server.
+  BasePath string
```

Comment:
I brought this up because our own main apiserver has different addresses inside and outside of the cluster, and when it serves discovery to an outside the cluster client, it needs to give them an address for 3rd party apiservers that works outside the cluster (i.e., not 10.x.x.x).